controllers/datafoundation_client.go

```diff
@@ -0,0 +1,241 @@
+			"an error occurred while parsing ManagedFusionOffering Data Foundation Client spec",
+		)
+		isValid = false
+	}
+	if dfStorageProviderEndpointAsURL.Host == "" {
+		r.Log.Error(
+			fmt.Errorf("cannot retrieve host from %s, might be a malformed or partial URL", dfStorageProviderEndpoint),
+			"an error occurred while parsing ManagedFusionOffering Data Foundation Client spec",
+		)
+		isValid = false
+	}
+
+	if !isValid {
+		r.Log.Info("parsing ManagedFusionOffering Data Foundation Client spec failed")
+		return fmt.Errorf("invalid ManagedFusionOffering Data Foundation Client spec")
+	}
+	r.Log.Info("parsing ManagedFusionOffering Data Foundation Client spec completed successfuly")
+
+	r.dataFoundationClientSpec = dataFoundationClientSpec{
+		onboardingTicket:          onboardingTicket,
+		dfStorageProviderEndpoint: dfStorageProviderEndpoint,
+	}
+	return nil
+}
+
+func (r *dataFoundationClientReconciler) reconcilePhases() (ctrl.Result, error) {
+	if !r.offering.DeletionTimestamp.IsZero() {
+		return ctrl.Result{}, nil
+	} else {
+		if err := r.reconcileStorageClient(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcileCSV(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcileDefaultBlockStorageClassClaim(); err != nil {
+			return ctrl.Result{}, err
+		}
+		if err := r.reconcileDefaultFSStorageClassClaim(); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+}
+
+func (r *dataFoundationClientReconciler) reconcileStorageClient() error {
+	r.Log.Info("Reconciling StorageClient")
+
+	_, err := r.CreateOrUpdate(&r.storageClient, func() error {
+		if err := r.own(&r.storageClient, true); err != nil {
+			return err
+		}
+		r.storageClient.Spec.OnboardingTicket = r.dataFoundationClientSpec.onboardingTicket
+		r.storageClient.Spec.StorageProviderEndpoint = r.dataFoundationClientSpec.dfStorageProviderEndpoint
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("unable to create/update StorageClient: %v", err)
+	}
+	return nil
+}
+
+func (r *dataFoundationClientReconciler) reconcileCSV() error {
+	r.Log.Info("Reconciling OCS Client Operator CSV")
+
+	csv, err := r.getCSVByPrefix(ocsClientOperatorName)
+	if err != nil {
+		return fmt.Errorf("unable to find OCS client CSV: %v", err)
+	}
+	err = r.GetAndUpdate(csv, func() error {
+		if err := r.own(csv, false); err != nil {
+			return fmt.Errorf("unable to set ownerRef on ocs client csv: %v", err)
+		}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("failed to update OCS client CSV: %v", err)
+	}
+	return nil
+}
+
+func (r *dataFoundationClientReconciler) reconcileDefaultBlockStorageClassClaim() error {
+	r.Log.Info("Reconciling Default BlockPool StorageClassClaim")
+
+	_, err := r.CreateOrUpdate(&r.defaultBlockStorageClassClaim, func() error {
+		if err := r.own(&r.defaultBlockStorageClassClaim, true); err != nil {
+			return fmt.Errorf("unable to set ownerRef on Default Blockpool StorageClassClaim: %v", err)
```

Comment:
No need for a custom error here as `r.own` will provide an error with a context. This is the same practice we follow everywhere else we use `r.own`.
```suggestion
			return err
```

Reply:
upadated